manifests/foreman_certs.pp

```diff
@@ -1,45 +0,0 @@
-# @summary Class for setting up Foreman client certs
-# @api private
-class katello_devel::foreman_certs (
-  $hostname       = $certs::node_fqdn,
-  $generate       = $certs::generate,
-  $regenerate     = $certs::regenerate,
-  $deploy         = $certs::deploy,
-) {
-
-  $client_cert_name = "${hostname}-foreman-client"
-  $client_cert      = "${certs::pki_dir}/certs/${client_cert_name}.crt"
-  $client_key       = "${certs::pki_dir}/private/${client_cert_name}.key"
-  $ssl_ca_cert      = $certs::ca_cert
-
-  # cert for authentication of puppetmaster against foreman
-  cert { $client_cert_name:
-    hostname      => $hostname,
-    purpose       => client,
-    country       => $certs::country,
-    state         => $certs::state,
-    city          => $certs::city,
-    org           => 'FOREMAN',
-    org_unit      => 'PUPPET',
-    expiration    => $certs::expiration,
-    ca            => $certs::default_ca,
-    generate      => $generate,
-    regenerate    => $regenerate,
-    deploy        => $deploy,
-    password_file => $certs::ca_key_password_file,
-  }
-
-  Cert[$client_cert_name] ~>
-  pubkey { $client_cert:
-    key_pair => Cert[$client_cert_name],
-  } ~>
-  privkey { $client_key:
-    key_pair => Cert[$client_cert_name],
-  } ~>
-  file { $client_key:
-    group => $katello_devel::group,
```

Comment:
This will break, right? The group is different. What I think we should do is reuse [certs::foreman](https://github.com/theforeman/puppet-certs/blob/master/manifests/foreman.pp) but update it to accept a group parameter. Then we can use it like this:
```puppet
class { 'certs::foreman':
  group => $katello_devel::group,
}
```

Remember why I never want hardcoded values but always parameters, even if we don't immediately use them? this is why :)

Reply:
Seems like a pretty quick change to make. Waiting for resolution (one way or the other) before I test

Reply:
Wellll, there are a couple nuances to it we could go by here. So let me layout the current, and sorta two approaches.

##### Current

The development environment deploys the following for Foreman client certificates on a standard Vagrant environment:

```
/etc/pki/katello/certs/foreman.example.com-foreman-client.crt
/etc/pki/katello/private/foreman.example.com-foreman-client.key
```

Where the certificate is world readable and the key is `root:vagrant` with `440` permissions.

A production installation deploys as:

```
/etc/foreman/client_cert.pem
/etc/foreman/client_key.pem
```

With `root:foreman` ownership for both and `440` for both.

##### Solutions

Reminder, these are only for development environments so trying to balance matching production deployments as closely as possible with ease of development environments and discoverability.

Option 1

Allow a production like deployment of client certificates to `/etc/foreman` and set permissions to `root:vagrant`

Option 2

Allow a production like deployment of client certificates to `/etc/foreman` and put the vagrant user in the foreman group.

Option 3

Deploy client certificates to `/home/vagrant/foreman/config` with `vagrant:vagrant` ownership.

Reply:
I like (2) but would be fine with just (1) for now. Guessing there's a bit more work to be done in order to achieve (2)?

Reply:
I *think* 2 is easier because you just remove all this code, and use a virtual resource to add the user to the foreman group.

Reply:
Sounds good to me. Let's see what @ekohl thinks

Reply:
+1 on the user in the group

Reply:
Option 3 would require some updates to `.gitignore` in Foreman to specifically ignore the files we are laying down or a special directory inside `config/` to put temporary devel environment files (e.g. `config/certs` or `config/tls`). This could be nice from a more encapsulated deployment perspective.

Reply:
You can of course also place them in `/home/vagrant/foreman-ssl` and point to them. Perhaps that's even better since then users can wipe their git checkout without cleaning the certs.

Reply:
I like the idea of putting the certs in a more dev friendly location longer term. In the short term, to re-use production code and reduce scope in here, I think we'll need the owner/group setting (here https://github.com/theforeman/puppet-certs/pull/374) and then have this declare that and create `/etc/foreman`.

Reply:
You already have the parameters (`client_cert`, `client_key`, and `ssl_ca_cert`) so you can easily modify those. I'm not sure it really increases the scope too much.

Reply:
Because of the way https://github.com/theforeman/puppet-certs/blob/master/manifests/foreman.pp works, I would still have to ensure `/etc/foreman` exists, or set deploy to false and then implement something akin to https://github.com/theforeman/puppet-certs/pull/370. So my general point is, right now, to unblock devel its easier to abide by the production deployment model and then as we work through making puppet-certs more flexible, we consider adjusting as part of that how we deploy the devel env.

Reply:
Is that because the `cert` resource itself is putting something in `/etc/foreman`? Because I don't really see what else requires it if you pass the parameters.

Reply:
`privkey` and `pubkey` handle putting things into `/etc/foreman`. However, if I pass `deploy => false` to that resource, then the RPMs won't get installed that place the certificate and key into `/etc/pki/katello-certs-tools` which is what `privkey` and `pubkey` work based upon. So, that is why I need the solution of deploying from the build dir (https://github.com/theforeman/puppet-certs/pull/370) which I could use puppet-katello_devel to pilot that.

I want to unblock developers first, and then look to make changes to the environment. Thus I would like to go with this solution and follow up with the build directory deployment model.

Reply:
Full circle -- https://github.com/theforeman/puppet-katello_devel/issues/265

Reply:
For what it's worth: I ran https://github.com/ekohl/puppet-certs/commit/adec2d6dfb74a86bf76dfdd2f49d607f895817be locally and it passes. No `/etc/foreman` exists. IMHO there is no reason to create `/etc/foreman`. While I didn't test https://github.com/ekohl/puppet-katello_devel/commit/8e2f945cd4e70d8e39737ad5a90c3e431442f22b, based on the puppet-certs test it should work.

Reply:
That should in theory work, I'll give it a go, thanks!